### PR TITLE
WIP: op_get_random_values

### PR DESCRIPTION
**Pull request by chiefbiiko**
First shot at #2321

The CSPRNG in use is Rust's [`rand::rngs::ThreadRng`](https://docs.rs/rand/0.6.5/rand/rngs/struct.ThreadRng.html).

Exposed as `globalThis.crypto.getRandomValues()`.

Following `Crypto.getRandomValues()`, but not completely.

Overlaps:
  - function signature (passing an integer-based TypedArray)

Differences:
  - errors
    - not throwing a `QuotaExceedError` if the input length is `gt` 65536
      - now throwing an `AssertionError` for this case
    - ~~passing `null` as input does not raise an exception until reaching `Rust`land~~
  - ~~sync and `async` variations~~
  - ~~additional support for `BigInt64Array` and `BigUint64Array`~~

~~Since these differences are incompatible with the browers' `Crypto` interface I just polluted `Deno`... 4give~~

### CODE REVIEW

**Comment by zekth** [comment]
i personnaly think `Crypto.getRandomValues` has to be similar as the browser implementation. So maybe add `Crypto.getRandomValuesAsync()` ?

File: cli/main.rs [hunk 1]
```diff
@@ -11,6 +11,7 @@ extern crate clap;
 extern crate deno;
 #[cfg(unix)]
 extern crate nix;
+extern crate rand;
```
Comment (zekth) [inline cli/main.rs]: this is not necessary IMO.

File: js/get_random_values.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+export function getRandomValuesSync(
+  typedArray:
+    | Int8Array
+    | Uint8Array
+    | Int16Array
+    | Uint16Array
+    | Int32Array
+    | Uint32Array
+): void {
+  dispatch.sendSync(...req(typedArray));
```
Comment (ry) [inline js/get_random_values.ts]: Do sync only. No need for an async version.

File: js/deno.ts [hunk 3]
```diff
@@ -81,6 +81,7 @@ export {
 export { inspect } from "./console";
 export { build, platform, OperatingSystem, Arch } from "./build";
 export { version } from "./version";
+export { getRandomValues, getRandomValuesSync } from "./get_random_values";
```
Comment (ry) [inline js/deno.ts]: Have a look in js/globals.ts 

you can export symbols globally there.

File: js/get_random_values.ts (hunk 2 above)
Comment (chiefbiiko) [reply js/get_random_values.ts]: Done

File: js/deno.ts (hunk 3 above)
Comment (chiefbiiko) [reply js/deno.ts]: Done

**Comment by ry** [comment]
You need to add `get_random_values_test.ts` to `js/unit_tests.ts`

File: js/get_random_values.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+export function getRandomValues<
+  T extends
+    | Int8Array
+    | Uint8Array
+    | Uint8ClampedArray
+    | Int16Array
+    | Uint16Array
+    | Int32Array
+    | Uint32Array
```
Comment (ry) [inline js/get_random_values.ts]: In js/types.ts we have `TypedArray`

File: js/globals.ts [hunk 5]
```diff
@@ -135,3 +142,19 @@ export interface ImportMeta {
...
+  readonly subtle: null;
+  getRandomValues: <
+    T extends
+      | Int8Array
+      | Uint8Array
+      | Uint8ClampedArray
+      | Int16Array
+      | Uint16Array
+      | Int32Array
+      | Uint32Array
```
Comment (ry) [inline js/globals.ts]: Use TypedArray from js/types.ts

**Comment by ry** [comment]
Currently CI is failing due to formatting. If you look at the end of the logs you'll see this:
```
Run tools/format.py 
 M cli/ops.rs
```


Also be sure to run "tools/lint.py"

File: js/get_random_values.ts (hunk 4 above)
Comment (chiefbiiko) [reply js/get_random_values.ts]: But that `TypedArray` type is missing some integer-based typed array types and also encompasses `Float32Array` which browsers' `Crypto.getRandomValues()` don't allow.

File: cli/main.rs (hunk 1 above)
Comment (chiefbiiko) [reply cli/main.rs]: My `rustc` demands it.

File: js/get_random_values.ts (hunk 4 above)
Comment (ry) [reply js/get_random_values.ts]: Can you add them?

Comment (ry) [reply js/get_random_values.ts]: Ah ok

**Pull request merged**